viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -4,11 +4,29 @@ import { TraceServerConfigService } from '../common/trace-server-config';
 @injectable()
 export class TraceServerServiceImpl implements TraceServerConfigService {
 
-    async startTraceServer(path: string | undefined, port: number | undefined): Promise<void> {
-        spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+    startTraceServer(path: string | undefined, port: number | undefined): Promise<void> {
+        const server = spawn(`${path}`, ['-vmargs', `-Dtraceserver.port=${port}`]);
+        return new Promise<void>((resolve, reject) => {
+            // If the server provides output on any channel before it exits, consider that a success.
+            server.stdout.on('data', () => resolve());
+            server.stderr.on('data', () => resolve());
+            // If the server exits or errors before it outputs, consider it a failure.
+            server.on('exit', code => reject(code));
+            server.on('error', error => reject(error));
+        })
+            .finally(() => { server.removeAllListeners(); });
```

Comment:
Here's a use case: another service that is not the trace server is running on port 8080. Data gets shown on stderr, so the promise resolve is called, but the server exits right after, but since you removed all listeners, it does not catch the exit. But since the server is never online, the timeout is hit and the message appears correctly after 8 seconds (instead of earlier as it could have been otherwise, fwiw here).

You can probably keep the server.on('exit') listener in general, so to detect the server process exiting at any moment (like a computer can decide to kill it if it takes too much RAM or CPU). If this happens without the user having asked for it, we can show her a message.

Reply:
I agree that that case deserves to be handled, but it can't be handled by the promise as currently structured. Since promises can only be rejected or resolve once - any call to `resolve` or `reject` thereafter is a no-op - once we have resolved or rejected the promise for any reason, no listeners can affect the state of the promise after that, and it's better to dispose of them. I suggest that the use case you mention be part of @ankusht-work's work on #226, which will include more detailed error handling for a wider range of cases. An additional argument for separating the two is that the server going down for any reason at any time is a problem that should be brought to the user's attention, and it it's worth having dedicated machinery outside of startup checks.

Reply:
However, if you'd like to see it handled here, one option would be to structure the promise logic so that we wait to resolve it on (seemingly) successful startup for a second or two so that the exit handler has time to catch startups with immediate termination before we resolve the promise.

Reply:
> > In short, I'd say, don't bother with the server listeners here, just spawn and wait for ping since that's how it's really making this work! But that's just my opinion
> 
> I'm also happy to take this route - I think you're right that ignoring the backend is fine for now, since it's the `ConnectionStateService` that plays the decisive role. On the other hand, the `error` listener does catch bad configurations earlier than waiting for a timeout.

I guess the real message in my comment is "don't keep code that doesn't really do what it's intended to do, like the resolves above". But if the rejects do have their utility, then yes keep them! And the port-in-use use case can indeed be handled by @ankusht-work 's patch. Sorry for the confusion.

Also, whatever you keep or don't keep, please add the proper comments in the code so we know how the status is actually handled, for instance, if  you need to keep the resolution, please say that resolution here does not mean the server is really started, the ping will do.

Reply:
@tahini, thanks for the feedback. I have added some detail to my comments, but left things otherwise unchanged. If we're interested in the possibility of rejection, we have to keep the resolution as well so the promise doesn't just hang.

Reply:
When I tried it this morning with gitpod, after clicking "Start trace server", 8 seconds later I got a toaster error "Failed to start server". Then after 22 seconds total, the status changed to "Stop trace server".

Reply:
So that problem can occur if the response to a bad ping is very slow. The `TraceServerConnectionStateService` only schedules a new ping when the last one is finished, whether it's successful or not. If a bad ping takes a long time to signal its failure, you can get into a no-man's land where you try to start the server (and it starts) during the time in which the `TraceServerConnectionStateService` is waiting on bad ping, and so doesn't make a good ping (which is usually much faster) until the time out has passed. One way to avoid this would be to timeout the pings in the `TraceServerConnectionStateService` so that it never waits too long before trying again. I'm happy to incorporate that fix into this PR, if you would like.

Reply:
I've modified the `TraceServerConnectionStatusService` with these lines:

```
const pingTimeout = new Promise((_, reject) => { setTimeout(reject, this.options.offlineTimeout); });
await Promise.race([this.tspClient.fetchExperiments(), pingTimeout]);
```

That should mean that any ping that takes more than the `offlineTimeout` (default of 5 seconds, it looks like) to settle (successfully or otherwise) will be considered a failure. That seems consistent with the intent of having an `offlineTimeout`, and should prevent the `Failed to start server.` message from ever appearing during a cycle during which the `TraceServerStatusService` would have signaled success - the timeout on the message waits two full cycles. It's still conceivable that the server might have started and simply take a long time to become ready to handle requests, but that contingency seems better left to the backend, or the entire strategy of depending on regular pinging to determine whether the server is active needs to be rethought.